lib/council-server/src/server.rs

```diff
@@ -68,19 +68,75 @@ impl Server {
             }
         });
 
+        // Before entering the main loop, tell everyone subscribing to the management channel (i.e. all pinga instances)
+        // to restart their jobs.
+        self.nats
+            .publish_with_reply(
+                management_channel.clone(),
+                management_reply_channel.clone(),
+                serde_json::to_vec(&Response::Restart)?.into(),
+            )
+            .await?;
+        info!(%management_channel, %management_reply_channel, "published message for all active pinga instances to restart jobs in progress");
+
+        // Begin the main loop. Everything after this point should be infallible.
+        self.core_loop_infallible_wrapper(
+            &mut subscriber,
+            &mut shutdown_request_rx,
+            &mut our_shutdown_request_rx,
+            &management_channel,
+            &management_reply_channel,
+        )
+        .await;
+
+        Ok(())
+    }
+
+    async fn core_loop_infallible_wrapper(
+        &self,
+        subscriber: &mut Subscriber,
+        shutdown_request_rx: &mut watch::Receiver<()>,
+        our_shutdown_request_rx: &mut watch::Receiver<()>,
+        management_channel: &ManagementChannel,
+        management_reply_channel: &ManagementReplyChannel,
+    ) {
         let mut complete_graph = ChangeSetGraph::default();
+        loop {
+            match self
+                .core_loop(
+                    subscriber,
+                    shutdown_request_rx,
+                    our_shutdown_request_rx,
+                    management_channel,
+                    management_reply_channel,
+                    &mut complete_graph,
+                )
+                .await
+            {
+                Ok(()) => break,
+                Err(e) => error!("{e}"),
+            }
+        }
```

Comment:
Keep looping until we have an explicit `Ok(())` return... which means "shutdown please!" or...
<img src="https://media4.giphy.com/media/sFoZicXyLjH7FnCVp2/giphy.gif"/>